guix: use GCC 14.3.0 over 13.3.0

fanquake opened the pull request:
Switching to using GCC 14.x for release builds has come up multiple times recently. It will eventually be needed for #25573, and could also be useful for #30210.

fanquake committed:
guix: use GCC 14.3.0 over 13.3.0

This will eventually be needed for #25573, and could be useful
for #30210.

fanquake committed:
ci: migrate some jobs to Debian Trixie, use GCC 14

hebasto commented:
Concept ACK.

maflcko reviewed (commented):
concept ack. 14.3 should be sufficiently propagated and tested by now, to rely on it for release builds

willcl-ark commented:
```
src/core/bitcoin on  pr-33775 [$?] via △ v3.31.6 via 🐍 v3.13.5 via ❄️  impure (nix-shell-env)
❯ uname -m; find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
x86_64
6c31a0f19daec94c3d629e50d6bc2461ceb6e0fbb0823cb10a40bc735e37f9b5  guix-build-25ede968bada/output/aarch64-linux-gnu/SHA256SUMS.part
f3ba2f0c5f1d99d377ea11ea2b9983187192d733c7db444060d112d1be48ecd2  guix-build-25ede968bada/output/aarch64-linux-gnu/bitcoin-25ede968bada-aarch64-linux-gnu-debug.tar.gz
321db43f3a4327287813274b7f9874fb76799643fdc9cbbc9ee13bce10b6e004  guix-build-25ede968bada/output/aarch64-linux-gnu/bitcoin-25ede968bada-aarch64-linux-gnu.tar.gz
58f449098f98b7746c2ba9286a3582852aed3c32b3cc0327598c279d46eab5b7  guix-build-25ede968bada/output/arm-linux-gnueabihf/SHA256SUMS.part
0c51f7872d5f65ad2bd432d20f76c49b8e2dc765cacb53f696b0a0c5ea91b56b  guix-build-25ede968bada/output/arm-linux-gnueabihf/bitcoin-25ede968bada-arm-linux-gnueabihf-debug.tar.gz
4f679e6074c745b2e91494721a6c8aa0ef24e19f18a3971ac53913441fdbf399  guix-build-25ede968bada/output/arm-linux-gnueabihf/bitcoin-25ede968bada-arm-linux-gnueabihf.tar.gz
8a66a9d3180f8fa386a0fdfc4ed1b80c716ec9b754921c35543cdc822ea35bf9  guix-build-25ede968bada/output/arm64-apple-darwin/SHA256SUMS.part
cbbffbb237ca4a0174f457824a21412334d7ce7c1931c78fe293520029e97c32  guix-build-25ede968bada/output/arm64-apple-darwin/bitcoin-25ede968bada-arm64-apple-darwin-codesigning.tar.gz
518312fec80b309a38393a7408cad0fda28f4cf7b52d08a5b4c2d1a26ea9b8d4  guix-build-25ede968bada/output/arm64-apple-darwin/bitcoin-25ede968bada-arm64-apple-darwin-unsigned.tar.gz
5a36a5ddea0070e3efd6e3db92528705483a2dd4ebc9287ed42c3d180537092a  guix-build-25ede968bada/output/arm64-apple-darwin/bitcoin-25ede968bada-arm64-apple-darwin-unsigned.zip
768446c9e4037d07b76df5379ff42482e2ce264db5e0ed61694af0c7ea49cd69  guix-build-25ede968bada/output/dist-archive/bitcoin-25ede968bada.tar.gz
a85baf46acec900f79b9343f9bd918731bfc0f1fea7586b902f28e99cda59e37  guix-build-25ede968bada/output/powerpc64-linux-gnu/SHA256SUMS.part
c01ac68acbdae911b56a6b3eb1762d9f57d4d7fc0dbedfa3f4c0905bf4be6853  guix-build-25ede968bada/output/powerpc64-linux-gnu/bitcoin-25ede968bada-powerpc64-linux-gnu-debug.tar.gz
f08b8d0909eb6f6a7f4b483f338655b0c44ee85c219360b62c99165df5aed157  guix-build-25ede968bada/output/powerpc64-linux-gnu/bitcoin-25ede968bada-powerpc64-linux-gnu.tar.gz
a2e1c4960965d6315e3c89579b7224970229d01f1c9b60ff8ec1698a1bebb884  guix-build-25ede968bada/output/riscv64-linux-gnu/SHA256SUMS.part
6b4958341f9648ddaf7c57a788db7b229d6be4cc7f9548f3acd84c2ce990909a  guix-build-25ede968bada/output/riscv64-linux-gnu/bitcoin-25ede968bada-riscv64-linux-gnu-debug.tar.gz
2f03dcd2941388cecb4c7ae1d8e0b925e0bcc1c0f79f6676906b88ed3f98c1e0  guix-build-25ede968bada/output/riscv64-linux-gnu/bitcoin-25ede968bada-riscv64-linux-gnu.tar.gz
a8b021adae39e2f9beaa37e6fa370eaa6fac033505cdfe8fdfcb01f6b6786ce1  guix-build-25ede968bada/output/x86_64-apple-darwin/SHA256SUMS.part
10cd32f7b51cca2473d4c1a92cde51914ec36e5d90d46c623047b0bd4a6171ce  guix-build-25ede968bada/output/x86_64-apple-darwin/bitcoin-25ede968bada-x86_64-apple-darwin-codesigning.tar.gz
dfde14c907e39b547681600cf0a1056f88d4e26d67bda66bb0ebdacaa77755bb  guix-build-25ede968bada/output/x86_64-apple-darwin/bitcoin-25ede968bada-x86_64-apple-darwin-unsigned.tar.gz
3a69e45b82a8397d4638ca3cf08d436f7fffa3ebab4a664ae5785b0e02726ed5  guix-build-25ede968bada/output/x86_64-apple-darwin/bitcoin-25ede968bada-x86_64-apple-darwin-unsigned.zip
f490dc45e3401815c252e4f818d266a6076e225e17f8a83a64d0919f2ca419fd  guix-build-25ede968bada/output/x86_64-linux-gnu/SHA256SUMS.part
de096f32dd1396195723db299ae8f299f9a8a425f45a79f9414184b4ec98c252  guix-build-25ede968bada/output/x86_64-linux-gnu/bitcoin-25ede968bada-x86_64-linux-gnu-debug.tar.gz
1f035fbdcda7e22390b4dc89dfc765f98a01f95d3f39faf8a34be51c6dd4bb9f  guix-build-25ede968bada/output/x86_64-linux-gnu/bitcoin-25ede968bada-x86_64-linux-gnu.tar.gz
50437f34b662a2cf0dc55c6d54b8a74b19d5fcc02380490e1a5a6347a4c2520c  guix-build-25ede968bada/output/x86_64-w64-mingw32/SHA256SUMS.part
780df1f42e33f6d5203b818bc0081c7a420878de274295cebb4668f54653a735  guix-build-25ede968bada/output/x86_64-w64-mingw32/bitcoin-25ede968bada-win64-codesigning.tar.gz
8eec58e81a5a3f072c8ba2ec39cde884f725e0ce3a295dd8f0b37710c0874682  guix-build-25ede968bada/output/x86_64-w64-mingw32/bitcoin-25ede968bada-win64-debug.zip
b7d9273ef4d2edf723442bedc570af20efa218afcd7cc85cfd66be065c3f6e58  guix-build-25ede968bada/output/x86_64-w64-mingw32/bitcoin-25ede968bada-win64-setup-unsigned.exe
68a88570939913939f411cfd11e0f0d3260970c4767490b227d94a417f9a55db  guix-build-25ede968bada/output/x86_64-w64-mingw32/bitcoin-25ede968bada-win64-unsigned.zip
```

willcl-ark commented:
Without the new flags I get:

```
core/worktrees/pr-33775 on  pr-33775 [$] via △ v3.31.6 via 🐍 v3.13.5 via ❄️  impure (nix-shell-env) took 3h44m19s
❯ uname -m; find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
x86_64
36feb0fb0ca11e5c17e01704b2c0c2cf6e23f59a62f0dd1d76819db5dc9fc1a5  guix-build-fc564434b8fb/output/aarch64-linux-gnu/SHA256SUMS.part
1b88fb5d39140df4b14e67638f02bb07ff2d7525b7ad812f783b4c0f87531908  guix-build-fc564434b8fb/output/aarch64-linux-gnu/bitcoin-fc564434b8fb-aarch64-linux-gnu-debug.tar.gz
9175e16424510fb37ffd2a674aad4f7ef8eb639f260c076b518785a0f81ed3d6  guix-build-fc564434b8fb/output/aarch64-linux-gnu/bitcoin-fc564434b8fb-aarch64-linux-gnu.tar.gz
80efa6885b84ac84d103a48045cd179e8416dca998dee86a075558d1d596faaf  guix-build-fc564434b8fb/output/arm-linux-gnueabihf/SHA256SUMS.part
2ba5ea3e1189f916849020c791fa09b618f376c5aaa61cc729745deaf0347dce  guix-build-fc564434b8fb/output/arm-linux-gnueabihf/bitcoin-fc564434b8fb-arm-linux-gnueabihf-debug.tar.gz
fd7c6c6a26a190b8545ba1db2352003c87c78c6164b911de929235180a75d78d  guix-build-fc564434b8fb/output/arm-linux-gnueabihf/bitcoin-fc564434b8fb-arm-linux-gnueabihf.tar.gz
8d9a2b291a0b508a2abc1dcfdbf81eb2da06e1e3a6f45ae458152a2976fd3548  guix-build-fc564434b8fb/output/arm64-apple-darwin/SHA256SUMS.part
c2eff8b7576a0de4450e70689bc883f19bf72ce60e27098e3589e995fb7592ec  guix-build-fc564434b8fb/output/arm64-apple-darwin/bitcoin-fc564434b8fb-arm64-apple-darwin-codesigning.tar.gz
27795225bfb45932d84716078c36bd5e46cbbc8f4025a1b275becd138fff8d9e  guix-build-fc564434b8fb/output/arm64-apple-darwin/bitcoin-fc564434b8fb-arm64-apple-darwin-unsigned.tar.gz
dde173ad4365dee349e4cc0317c4eb370b18756841f2ce17fe870ec95525b1a9  guix-build-fc564434b8fb/output/arm64-apple-darwin/bitcoin-fc564434b8fb-arm64-apple-darwin-unsigned.zip
7cfaeaf3fdfdd6525f73c0b95945b497bc37a875a16c5e2f87b3c3db370b1463  guix-build-fc564434b8fb/output/dist-archive/bitcoin-fc564434b8fb.tar.gz
c9c6434a74c68b908f7ee2376017decdc2cdaafc7130efcd97ddd79a9ee30455  guix-build-fc564434b8fb/output/powerpc64-linux-gnu/SHA256SUMS.part
e21646e7d4f1e6feacffa5fd6d7c7932b1933b18c505683af4d3cd29f6cec5ed  guix-build-fc564434b8fb/output/powerpc64-linux-gnu/bitcoin-fc564434b8fb-powerpc64-linux-gnu-debug.tar.gz
c327f0d153b54be042805c7039909c1fb499b53c6865251f804a422ccce076d0  guix-build-fc564434b8fb/output/powerpc64-linux-gnu/bitcoin-fc564434b8fb-powerpc64-linux-gnu.tar.gz
17f6a43afeab0d604c622b6446cd0ce91941446583b0992dfe7de3ac373805a3  guix-build-fc564434b8fb/output/riscv64-linux-gnu/SHA256SUMS.part
c55582b7a41b420d469718c3863cc4f422919ed9a40889b2158b55a70f0ca242  guix-build-fc564434b8fb/output/riscv64-linux-gnu/bitcoin-fc564434b8fb-riscv64-linux-gnu-debug.tar.gz
59797ae9b27b5c4756d5b86ac3f96f42d3a516cdd51403302b990945e613f518  guix-build-fc564434b8fb/output/riscv64-linux-gnu/bitcoin-fc564434b8fb-riscv64-linux-gnu.tar.gz
a647c87cbaebfe51d2b53ca87e84e9fb7f0ca1c18d7f1470df57f4b3fb7974f3  guix-build-fc564434b8fb/output/x86_64-apple-darwin/SHA256SUMS.part
f33fb123db84d248aa50a228851dff5ec7fc3c49808b1b826bd3a6519f030432  guix-build-fc564434b8fb/output/x86_64-apple-darwin/bitcoin-fc564434b8fb-x86_64-apple-darwin-codesigning.tar.gz
da4d7d1c9f2cbe77e46ea776e58111fd31be998707b9c29886712616f91f3cbb  guix-build-fc564434b8fb/output/x86_64-apple-darwin/bitcoin-fc564434b8fb-x86_64-apple-darwin-unsigned.tar.gz
e35de680539a1763d25d6ac5134786b693aeb3386124d8bb768d2cca36d5308c  guix-build-fc564434b8fb/output/x86_64-apple-darwin/bitcoin-fc564434b8fb-x86_64-apple-darwin-unsigned.zip
ac1a14a98dfeac54139b556725970f9fc11050b866324c5c7b489582d30146d0  guix-build-fc564434b8fb/output/x86_64-linux-gnu/SHA256SUMS.part
df6e6062d4f4dc5c39eea4501beb7ea7c06fa286f3bc33905cdb550bb34464a4  guix-build-fc564434b8fb/output/x86_64-linux-gnu/bitcoin-fc564434b8fb-x86_64-linux-gnu-debug.tar.gz
084839fecaaccf3f674dd63dcb5ad467f9892568e72d9493f622c388cc2cd80b  guix-build-fc564434b8fb/output/x86_64-linux-gnu/bitcoin-fc564434b8fb-x86_64-linux-gnu.tar.gz
7731e0ebe19bb9f19491f87879debdc671dfc2c3b87de816577223ef95ab2b4d  guix-build-fc564434b8fb/output/x86_64-w64-mingw32/SHA256SUMS.part
66bce8603fd3a78b0f2b277352ddd81fe6e20a7468e5958ecedfa26d636f6938  guix-build-fc564434b8fb/output/x86_64-w64-mingw32/bitcoin-fc564434b8fb-win64-codesigning.tar.gz
1bc1e2341163ad8c9ac3c4e6cad4b8c53c445e3798f5e7ca2f2b8efe1956e3cf  guix-build-fc564434b8fb/output/x86_64-w64-mingw32/bitcoin-fc564434b8fb-win64-debug.zip
121ee4b1dd3cfb4406d851866130a6b4120a20ab16a142dd6b2fc68e9863b2bc  guix-build-fc564434b8fb/output/x86_64-w64-mingw32/bitcoin-fc564434b8fb-win64-setup-unsigned.exe
8b57593558909a896394c7110e15515956a41c514c47f0b6b194563ca48dbd3b  guix-build-fc564434b8fb/output/x86_64-w64-mingw32/bitcoin-fc564434b8fb-win64-unsigned.zip
```

fanquake committed:
guix: disable building libgomp in GCC

Prunes:
libgomp.a
libgomp.la
libgomp.so
libgomp.so.1
libgomp.so.1.0.0
libgomp.spec

fanquake committed:
guix: disable libquadmath in GCC

Prunes:
libquadmath.a
libquadmath.la
libquadmath.so
libquadmath.so.0
libquadmath.so.0.0.0

fanquake committed:
guix: build GCC with --enable-host-bind-now

fanquake committed:
guix: disable gprofng in GCC

laanwj commented:
> Seeing the following difference:

That's such a strange difference. Just register selection.

We should probably aim to create a minimum reproduction, to be able to report and troubleshoot this.

laanwj commented:
This is a minimum C++ file (independent from bitcoin core headers) that will reproduce the issue:
```c++
#include <filesystem>
#include <string>
#include <fstream>
#include <vector>

#include <cstring> // Unused, but removing this makes it determinstic

namespace fs {

using namespace std::filesystem;

// Dummy class. Using "path = std::filesystem::path" makes it deterministic.
class path : public std::filesystem::path
{
public:
    using std::filesystem::path::path;

    path(std::filesystem::path path) : std::filesystem::path::path(std::move(path)) {}
};

}

bool IsBDBFile(const fs::path& path)
{
    // Removing this condition makes it deterministic.
    if (!fs::exists(path)) return false;
    return true;
}

bool IsSQLiteFile(const fs::path& path)
{
    // Removing this condition makes it deterministic.
    if (!fs::exists(path)) return false;
    return true;
}

std::vector<std::pair<fs::path, std::string>> ListDatabases(const fs::path& wallet_dir)
{
    std::vector<std::pair<fs::path, std::string>> paths;
    std::error_code ec;

    for (auto it = fs::recursive_directory_iterator(wallet_dir, ec); it != fs::recursive_directory_iterator(); it.increment(ec)) {
        const fs::path path{it->path().lexically_relative(wallet_dir)};

        if (it->status().type() == fs::file_type::directory) {
            if (IsBDBFile(it->path())) {
                paths.emplace_back(path, "bdb");
            } else if (IsSQLiteFile(it->path())) {
                paths.emplace_back(path, "sqlite");
            }
        } else if (it.depth() == 0 && it->symlink_status().type() == fs::file_type::regular && it->path().extension() != ".bak") {
            if (it->path().filename() == "wallet.dat") {
                if (IsBDBFile(it->path())) {
                    paths.emplace_back(fs::path(), "bdb");
                } else if (IsSQLiteFile(it->path())) {
                    paths.emplace_back(fs::path(), "sqlite");
                }
            } else if (IsBDBFile(it->path())) {
                paths.emplace_back(path, "bdb");
            }
        }
    }

    return paths;
}
```

To build, i'm currently carefully reproducing the conditions of Bitcoin Core's build system. i'm not sure in how far this is necessary, but remapping the `/gnu/store` prefixes is important, otherwise the platforms output differs.

```bash
# CFLAGS
HOST_CFLAGS="-O2 -g"
HOST_CFLAGS+=$(find /gnu/store -maxdepth 1 -mindepth 1 -type d -exec echo -n " -ffile-prefix-map={}=/usr" \;)
case "$HOST" in
    *mingw*)  HOST_CFLAGS+=" -fno-ident" ;;
    *darwin*) unset HOST_CFLAGS ;;
esac

# CXXFLAGS
HOST_CXXFLAGS="$HOST_CFLAGS"

EXTRA_CXXFLAGS="-fdump-tree-all"

OUT_OBJ=test.cpp.obj
cd /distsrc-base/distsrc-c4da80608710-x86_64-w64-mingw32/build && \
    x86_64-w64-mingw32-g++ -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DNOMINMAX -DWIN32 -DWIN32_LEAN_AND_MEAN -D_MT -D_WIN32_IE=0x0A00 -D_WIN32_WINNT=0x0A00 -D_WINDOWS \
    ${HOST_CXXFLAGS} \
    ${EXTRA_CXXFLAGS} \
    -O2 -O2 -g -fvisibility=hidden -Wa,-muse-unaligned-vector-move -fno-extended-identifiers -fdebug-prefix-map=/distsrc-base/distsrc-c4da80608710-x86_64-w64-mingw32/src=. -fmacro-prefix-map=/distsrc-base/distsrc-c4da80608710-x86_64-w64-mingw32/src=. -fstack-reuse=none -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -Wstack-protector -fstack-protector-all -fcf-protection=full -Wall -Wextra -Wformat -Wformat-security -Wvla -Wredundant-decls -Wdate-time -Wduplicated-branches -Wduplicated-cond -Wlogical-op -Woverloaded-virtual -Wsuggest-override -Wimplicit-fallthrough -Wunreachable-code -Wbidi-chars=any -Wundef -Wno-unused-parameter -std=c++20 \
    -MD -MT ${OUT_OBJ} -MF ${OUT_OBJ} -o ${OUT_OBJ} -c \
    /bitcoin/test.cpp && \
    sha256sum ${OUT_OBJ}
```

These are critical, without them, the non-determinism goes away:
- `#include <cstring>`
- Defining and using `fs::path` 
- `IsBDBFile` and `IsSQLiteFile` are defined and used, and have some logic in them

i think these are purely to prevent the compiler from inlining `std::vector<std::pair<fs::path, std::string>>::emplace_back`, which would cause the logic that causes the divergence in `objsz1` to not be hit.

fanquake commented:
@laanwj nice! I've re-integrated this into Guix (to build with `./contrib/guix/guix-build`), and minified the flags further, see this branch: https://github.com/fanquake/bitcoin/tree/repro_33775_minimal. With the top commit, the diff is still the register swapping:
```diff
--- a.txt
+++ b.txt
@@ -1,9 +1,9 @@
 
-test.cpp.obj.aarch64:     file format pe-x86-64
+test.cpp.obj.x86_64:     file format pe-x86-64
 
 
 Disassembly of section .text:
 
 0000000000000000 <IsBDBFile(fs::path const&)>:
        0:	48 83 ec 28          	sub    $0x28,%rsp
        4:	e8 00 00 00 00       	call   9 <IsBDBFile(fs::path const&)+0x9>
@@ -223,18 +223,18 @@
      332:	48 39 eb             	cmp    %rbp,%rbx
      335:	48 0f 42 dd          	cmovb  %rbp,%rbx
      339:	49 83 c7 48          	add    $0x48,%r15
      33d:	49 83 c4 48          	add    $0x48,%r12
      341:	48 8d 74 1e b8       	lea    -0x48(%rsi,%rbx,1),%rsi
      346:	4c 39 f7             	cmp    %r14,%rdi
      349:	0f 84 51 01 00 00    	je     4a0 <std::pair<fs::path, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >& std::vector<std::pair<fs::path, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<fs::path, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::emplace_back<fs::path const&, char const (&) [4]>(fs::path const&, char const (&) [4]) [clone .isra.0]+0x460>
-     34f:	49 39 f5             	cmp    %rsi,%r13
-     352:	48 89 f5             	mov    %rsi,%rbp
+     34f:	4c 39 ee             	cmp    %r13,%rsi
+     352:	4c 89 ed             	mov    %r13,%rbp
      355:	49 8d 4f 10          	lea    0x10(%r15),%rcx
-     359:	49 0f 43 ed          	cmovae %r13,%rbp
+     359:	48 0f 43 ee          	cmovae %rsi,%rbp
      35d:	49 89 0f             	mov    %rcx,(%r15)
      360:	49 8b 06             	mov    (%r14),%rax
      363:	48 89 eb             	mov    %rbp,%rbx
      366:	48 29 f3             	sub    %rsi,%rbx
      369:	48 29 ee             	sub    %rbp,%rsi
      36c:	4c 39 e0             	cmp    %r12,%rax
      36f:	0f 85 23 ff ff ff    	jne    298 <std::pair<fs::path, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >& std::vector<std::pair<fs::path, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<fs::path, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::emplace_back<fs::path const&, char const (&) [4]>(fs::path const&, char const (&) [4]) [clone .isra.0]+0x258>
```

laanwj commented:
Here's a patch to `manifest.scm` that makes it easier to test with arbitrary gcc commits: https://github.com/laanwj/bitcoin/commit/a323ddaf5ad88b3429e2ee079566efee4e0d8f7e

Allows:
```
GCC_COMMIT=953536524a198efa932ee2cccf5d29d4fc1d24f3 GCC_HASH=3b59e003321ac50462f626bba07826c2d4c203d4fd1500c64d2ea332e61eaab0 HOSTS="x86_64-w64-mingw32" contrib/guix/guix-build
```
- GCC_COMMIT is any arbitrary commit from the gcc repository (any will work as long as the patches apply).
- GCC_HASH is the hex sha256sum of the tar.gz file for the commit. Sadly guix won't work without a hash, but unlike the base32 ones this one is easy to pre-generate with stock utilities.

FWIW i've checked
```
bb24b4c804f3d95b0ba95b74965bd04f425f8467  (releases/gcc-14.3.0) bad [as expected]
c035a7c30c310ff928988cbcf445f3f21be10aa1  (releases/gcc-14.2.0) bad
ed10445fe222d3973ae13eda9bf211f315c5e3f9  (releases/gcc-14.1.0) bad
953536524a198efa932ee2cccf5d29d4fc1d24f3  (releases/gcc-13.3.0) good [as expected]
```
bisecting wlll go back to an ancient merge base (8e08c7886eed5824bebd0e011526ec302d622844) that breaks guix's patching, sadly.

It might be possible to cut down the toolchain (reduce the number of stages) for faster builds. Especially if we pre-preprocess test.cpp. But i haven't looked into this yet.

laanwj commented:
FWIW: My bisect is not done yet (2 steps to go, it will likely finish today), but in the current commit range https://github.com/gcc-mirror/gcc/commit/1d82fc2e6824bf83159389729c31a942f7b91b04 `optimize std::vector::push_back` is the most suspicious as it affects vector appending. It's a change to `libstdc++-v3`, not the compiler, so it doesn't directly point at a root cause.

laanwj committed:
guix: Apply SSA generation patch to maintain determinism

See:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=123351
https://gcc.gnu.org/pipermail/gcc-patches/2026-January/704817.html

hebasto reviewed (commented):
My Guix build:
```
aarch64
02de261cf6c90f15a69d42b43e63485a6c0cdb3d745b5916b6834278a1d63dbc  guix-build-a72ff31b879f/output/aarch64-linux-gnu/SHA256SUMS.part
cf2d9615338be34f5082940d7d2b1a4c085226a410a75dceec764e42cd5df5ee  guix-build-a72ff31b879f/output/aarch64-linux-gnu/bitcoin-a72ff31b879f-aarch64-linux-gnu-debug.tar.gz
e18f34ce40036367f611451ae4dfe39e7dedeb53e549f229cff118bf2c2836a8  guix-build-a72ff31b879f/output/aarch64-linux-gnu/bitcoin-a72ff31b879f-aarch64-linux-gnu.tar.gz
17cd0a5fad2a4e648128a7a5fce4b46a894153b45ca96bb1c7ad874c5946bc5d  guix-build-a72ff31b879f/output/arm-linux-gnueabihf/SHA256SUMS.part
dd2f24c67be7635f13e250594d82d78b22cb9414d89e447f41e8927b98d15b34  guix-build-a72ff31b879f/output/arm-linux-gnueabihf/bitcoin-a72ff31b879f-arm-linux-gnueabihf-debug.tar.gz
b2f6d967da6402885fc4a7271da9996c90350d02fc9395c7660191f5f855bb46  guix-build-a72ff31b879f/output/arm-linux-gnueabihf/bitcoin-a72ff31b879f-arm-linux-gnueabihf.tar.gz
63d80dd6a617ace6ce43ae11fc4c388eaf809341fba7304c76ed8d30a7608e0b  guix-build-a72ff31b879f/output/arm64-apple-darwin/SHA256SUMS.part
148af422c55455b91f9e7947bde53546aca9705eaedae4240d7eb439425c4c8d  guix-build-a72ff31b879f/output/arm64-apple-darwin/bitcoin-a72ff31b879f-arm64-apple-darwin-codesigning.tar.gz
8d5056542c7e47d4674328f65e05c4e8b871b823c33132c949eafd36a7dc5b7c  guix-build-a72ff31b879f/output/arm64-apple-darwin/bitcoin-a72ff31b879f-arm64-apple-darwin-unsigned.tar.gz
5a9748970538e1bf99a1c14ef0acab693e238ec212db6a3531d6743c117de725  guix-build-a72ff31b879f/output/arm64-apple-darwin/bitcoin-a72ff31b879f-arm64-apple-darwin-unsigned.zip
094363086ca8986f8460c8355abd7ff4f7b2e3dbdafab2a0c68c51e93b432e79  guix-build-a72ff31b879f/output/dist-archive/bitcoin-a72ff31b879f.tar.gz
a8bd7171de3fd69925f55cd24a3d76da19e7b6f11cc01ee101e29f81eb0f9df1  guix-build-a72ff31b879f/output/powerpc64-linux-gnu/SHA256SUMS.part
06ef995f52f14fd1b780f297898c41232c10b64931b2053622228642bf04afc5  guix-build-a72ff31b879f/output/powerpc64-linux-gnu/bitcoin-a72ff31b879f-powerpc64-linux-gnu-debug.tar.gz
c71593027c636481c7ec706947876c0086baa58e43170a8d4871568db6689234  guix-build-a72ff31b879f/output/powerpc64-linux-gnu/bitcoin-a72ff31b879f-powerpc64-linux-gnu.tar.gz
c44e49c68906657e90c328b8b05ee76d92bb6dc3280d13d0f82521cf66529c51  guix-build-a72ff31b879f/output/riscv64-linux-gnu/SHA256SUMS.part
576f0f27c4ea2ff720e548b502f265efaeebc0896c20d2ba40839728710bae8a  guix-build-a72ff31b879f/output/riscv64-linux-gnu/bitcoin-a72ff31b879f-riscv64-linux-gnu-debug.tar.gz
403c010fc0f9363eb6536875b9a6181ef835eb86b8ae538f9fd9430bd67e93f6  guix-build-a72ff31b879f/output/riscv64-linux-gnu/bitcoin-a72ff31b879f-riscv64-linux-gnu.tar.gz
8c6f83e6686cb3e486568b7353a08562e361b3abd4281ad08c47d0e74d071ea8  guix-build-a72ff31b879f/output/x86_64-apple-darwin/SHA256SUMS.part
8271c2bd3c6d4dc17ce6eb316e4094c759b8be15fd3b308ccd6856ceb0a09b86  guix-build-a72ff31b879f/output/x86_64-apple-darwin/bitcoin-a72ff31b879f-x86_64-apple-darwin-codesigning.tar.gz
c70b4b4a3a23ca9ffb442d00257f6bf8233792785e4a3d391be132cceaec1e81  guix-build-a72ff31b879f/output/x86_64-apple-darwin/bitcoin-a72ff31b879f-x86_64-apple-darwin-unsigned.tar.gz
3abbf804472200486f757181c44e7e19aeb8119f825a6580d735d98e144ee2cd  guix-build-a72ff31b879f/output/x86_64-apple-darwin/bitcoin-a72ff31b879f-x86_64-apple-darwin-unsigned.zip
01b1696b94b9f22531c364a0c3f30a9064a5b2baff59a8333a0b38aa19be2a7b  guix-build-a72ff31b879f/output/x86_64-linux-gnu/SHA256SUMS.part
86acfda0a08f9b4eddab48bf4cd723c9015b5b2d84bffecdf8175df57dac5051  guix-build-a72ff31b879f/output/x86_64-linux-gnu/bitcoin-a72ff31b879f-x86_64-linux-gnu-debug.tar.gz
30a3c4ef58e4bef72c712631c7b61099e2da8391b06926863a7c2003ca8dc97b  guix-build-a72ff31b879f/output/x86_64-linux-gnu/bitcoin-a72ff31b879f-x86_64-linux-gnu.tar.gz
6fafb7b8801eb37d37abd1ea4907e6c93bbac06f2fc6b59d4cd771b38ba97e89  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/SHA256SUMS.part
3e214aacaf379fefead8694b1962c4f4c38b2754eae7069de6bfb259ad80e3c3  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-codesigning.tar.gz
fb3f9479336034177b9309390f667047c833fb1e944dbcccbd6a097e6988de8c  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-debug.zip
b0ff4b5811fca1bc654e35d77828a4b025fe756e71314b6f91501f27375bd81a  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-setup-unsigned.exe
d7a2204da3f11afcd9851804fb56132d596cb5157c1a558045eb2b2689a6259f  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-unsigned.zip
```

UPDATE: Same hashes on `x86_64`.

hebasto reviewed (commented):
Approach ACK a72ff31b879f164ca7875c0121953af93d799aee.

hebasto reviewed (commented):
My partial Guix build on RISC-V:
```
riscv64
094363086ca8986f8460c8355abd7ff4f7b2e3dbdafab2a0c68c51e93b432e79  guix-build-a72ff31b879f/output/dist-archive/bitcoin-a72ff31b879f.tar.gz
6fafb7b8801eb37d37abd1ea4907e6c93bbac06f2fc6b59d4cd771b38ba97e89  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/SHA256SUMS.part
3e214aacaf379fefead8694b1962c4f4c38b2754eae7069de6bfb259ad80e3c3  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-codesigning.tar.gz
fb3f9479336034177b9309390f667047c833fb1e944dbcccbd6a097e6988de8c  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-debug.zip
b0ff4b5811fca1bc654e35d77828a4b025fe756e71314b6f91501f27375bd81a  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-setup-unsigned.exe
d7a2204da3f11afcd9851804fb56132d596cb5157c1a558045eb2b2689a6259f  guix-build-a72ff31b879f/output/x86_64-w64-mingw32/bitcoin-a72ff31b879f-win64-unsigned.zip
```

hebasto reviewed (commented):
In 92f275aabf12ebfc0d6d0ccbd83af7e4f88dad25:

From the partially reverted [commit message](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=1d82fc2e6824bf83159389729c31a942f7b91b04), it follows that the commit affects code generation at `-O3`. This is somewhat confusing, as we use `-O2`.

Based on that commit message, I conclude that reverting only the two checks, rather than the entire commit, is safe. Nevertheless, why not revert the whole commit, which would make review of this PR much easier?

maflcko commented:
> Based on that commit message, I conclude that reverting only the two checks, rather than the entire commit, is safe. Nevertheless, why not revert the whole commit, which would make review of this PR much easier?

I'd say it is easier to check that dead/unreachable code has been removed than to check the correct port of a optimization reversion across gcc minor or even major versions?

fanquake commented on contrib/guix/patches/gcc-revert-optimize-std-vector-push_back.patch:
```diff
@@ -0,0 +1,38 @@
+Subject: [PATCH] Revert builtin_unreachable introduced in commit
+ 1d82fc2e6824bf83159389729c31a942f7b91b04.
+
+For some reason, these make the build non-deterministic between
```
Nice. Pulled into the patch.

hebasto reviewed (commented):
My Guix build:
```
x86_64
40ad33c821da20854bd5c01127e0aeec4b387ca44c485655bfadc0c65a80e9ce  guix-build-2a746500fa76/output/aarch64-linux-gnu/SHA256SUMS.part
fe01b5d9c7619a9f8b01cccad0655c9dfe3703ea053ff9438c6941d6e8b13c78  guix-build-2a746500fa76/output/aarch64-linux-gnu/bitcoin-2a746500fa76-aarch64-linux-gnu-debug.tar.gz
8d12bb46c0e4a6b178c821817250cdd00e6b16fa2e77e129b9224973e9ad0445  guix-build-2a746500fa76/output/aarch64-linux-gnu/bitcoin-2a746500fa76-aarch64-linux-gnu.tar.gz
600393616d9275202807b2f83a966f55f47d7313d9b9291587e3a4e9f23b6181  guix-build-2a746500fa76/output/arm-linux-gnueabihf/SHA256SUMS.part
9c75f854b3f3f92dd81081af789f32b60661fd0f5d1bd9973ccd7c3cff884196  guix-build-2a746500fa76/output/arm-linux-gnueabihf/bitcoin-2a746500fa76-arm-linux-gnueabihf-debug.tar.gz
08ac981dd5d5fb4dd943f0ea93a6281387a46b3c42c9b06a7ff0c99e39868f65  guix-build-2a746500fa76/output/arm-linux-gnueabihf/bitcoin-2a746500fa76-arm-linux-gnueabihf.tar.gz
2dbdffc0dcde53df241d8d67e1cce6a155425c3de55f08fc3483aa62181cfae9  guix-build-2a746500fa76/output/arm64-apple-darwin/SHA256SUMS.part
50e4268ff7039ca10c6b21b58322513c455a28477afeb61240a178ea23a72ab9  guix-build-2a746500fa76/output/arm64-apple-darwin/bitcoin-2a746500fa76-arm64-apple-darwin-codesigning.tar.gz
8137daa7a41a08279382fb63f2030802ddc538a32fe0cedae6db8e7897d53f29  guix-build-2a746500fa76/output/arm64-apple-darwin/bitcoin-2a746500fa76-arm64-apple-darwin-unsigned.tar.gz
2994d09b304cc0860b1b92553715933f3a54d53555cb60dcb6ebf57c2dc58a6e  guix-build-2a746500fa76/output/arm64-apple-darwin/bitcoin-2a746500fa76-arm64-apple-darwin-unsigned.zip
2af7c1cb984b6eec476ff441e04fcc3f5a7d26f88a9e00f8773f08cd74e88a7b  guix-build-2a746500fa76/output/dist-archive/bitcoin-2a746500fa76.tar.gz
dac539f42ea8115867ac3693e925f86c5ee7fbda7374e045f8c01473fceafb41  guix-build-2a746500fa76/output/powerpc64-linux-gnu/SHA256SUMS.part
1b66865391c52292a0739a6fa69024bf850c6b41f0768d4e9db045aeffc4c5c8  guix-build-2a746500fa76/output/powerpc64-linux-gnu/bitcoin-2a746500fa76-powerpc64-linux-gnu-debug.tar.gz
1faa4b649573ead908d51a43a78298ae0cd2015e6054bd8790d19acbdde74a5d  guix-build-2a746500fa76/output/powerpc64-linux-gnu/bitcoin-2a746500fa76-powerpc64-linux-gnu.tar.gz
0388e0e34f28e1f65c1a0315603a8664529996871e062aa53bc10657663353c0  guix-build-2a746500fa76/output/riscv64-linux-gnu/SHA256SUMS.part
0f09706414d5a15456e146819aaefd08783c8f0732a4c38bfe105bb9d50387d0  guix-build-2a746500fa76/output/riscv64-linux-gnu/bitcoin-2a746500fa76-riscv64-linux-gnu-debug.tar.gz
0b3ef8bb5ec28d888f3c37ccf1ae95312fb4645c4c55770c95073091abe7fde4  guix-build-2a746500fa76/output/riscv64-linux-gnu/bitcoin-2a746500fa76-riscv64-linux-gnu.tar.gz
bcbb8f2a8e79874a6142e4eb605cf591a331f6a0448eac1801fe01996df22835  guix-build-2a746500fa76/output/x86_64-apple-darwin/SHA256SUMS.part
3d3abd9904e286562c22a7ab10bb992aa837b37554a4c151b7051bf2cdbaf483  guix-build-2a746500fa76/output/x86_64-apple-darwin/bitcoin-2a746500fa76-x86_64-apple-darwin-codesigning.tar.gz
63181d3518d3a913ce25bd3e220e8385701a76891209431ed6b6efe4e419a9bc  guix-build-2a746500fa76/output/x86_64-apple-darwin/bitcoin-2a746500fa76-x86_64-apple-darwin-unsigned.tar.gz
4ea713112c04b367d257676966cab796a2ff00208a393df2c8fc2f8a5af73f0a  guix-build-2a746500fa76/output/x86_64-apple-darwin/bitcoin-2a746500fa76-x86_64-apple-darwin-unsigned.zip
ace818bca8fc5041483b10ceeb9e1336fd26363781167decbc32c31cbb60fdd9  guix-build-2a746500fa76/output/x86_64-linux-gnu/SHA256SUMS.part
76628eb51a70ed7a47f7f8da72b414654a94a952f27ea0722d0625c5952bd505  guix-build-2a746500fa76/output/x86_64-linux-gnu/bitcoin-2a746500fa76-x86_64-linux-gnu-debug.tar.gz
2e65bee126f47702020d52c8c8f23871a39e121e983b5e5d6d04f5d9103dde57  guix-build-2a746500fa76/output/x86_64-linux-gnu/bitcoin-2a746500fa76-x86_64-linux-gnu.tar.gz
7b228241e4f8c8740424f9663b0fdfc362619897fdeb4eac69eee49aff4ebd45  guix-build-2a746500fa76/output/x86_64-w64-mingw32/SHA256SUMS.part
782fce8cfcaba112dc1633ee18a0d81a3a739186bbbd250303cb4393c4b5ba76  guix-build-2a746500fa76/output/x86_64-w64-mingw32/bitcoin-2a746500fa76-win64-codesigning.tar.gz
7b1da47aa555de00bf1f0720ad7464c2d3ccc40ebc257aa27fc1c02407d74ebd  guix-build-2a746500fa76/output/x86_64-w64-mingw32/bitcoin-2a746500fa76-win64-debug.zip
ed26ec5df9f62a063c6239698008b71acf82f6e0d7e3ca0a023f7e7ff2094427  guix-build-2a746500fa76/output/x86_64-w64-mingw32/bitcoin-2a746500fa76-win64-setup-unsigned.exe
356fad06147ecc21007445ab5be5c72ea3125877f907e5ebc072686b610e2b35  guix-build-2a746500fa76/output/x86_64-w64-mingw32/bitcoin-2a746500fa76-win64-unsigned.zip
```

UPDATE: Got the same hashes on `aarch64`.

hebasto reviewed (approved):
ACK 2a746500fa769f83405349bcffa3f0971b92a2de. I have reviewed the code and it looks OK. The new GCC patch looks reasonable.

theuni commented on contrib/guix/manifest.scm:
```diff
@@ -421,6 +421,7 @@ inspecting signatures in Mach-O binaries.")
             ;; https://gcc.gnu.org/install/configure.html
             (list "--enable-threads=posix",
                   "--enable-default-ssp=yes",
+                  "--enable-host-bind-now=yes",
```
For context:
> Specify that the host executables should be linked with the option -Wl,-z,now, which means that the dynamic linker will resolve all symbols when the executables are started, and that in turn allows RELRO to mark the GOT read-only, resulting in better security.

theuni commented on contrib/guix/manifest.scm:
```diff
@@ -422,6 +422,7 @@ inspecting signatures in Mach-O binaries.")
             (list "--enable-threads=posix",
                   "--enable-default-ssp=yes",
                   "--disable-gcov",
+                  "--disable-libgomp",
```
Are we sure that none of our dependencies use this anywhere? From a quick look, it doesn't seem like gcc itself does. And libstdc++ does, but only for the experimental `parallel/*` ts.

I'd hate to see us quietly losing performance somewhere. But as far as I can tell, none of our deps are using it at the moment.

theuni commented on contrib/guix/manifest.scm:
```diff
@@ -123,7 +123,7 @@ desirable for building Bitcoin Core release binaries."
 (define (gcc-libgcc-patches gcc)
   (package-with-extra-patches gcc
-    (search-our-patches "gcc-remap-guix-store.patch")))
+    (search-our-patches "gcc-remap-guix-store.patch" "gcc-ssa-generation.patch")))
```
Beautiful! Nice work @maflcko !

theuni reviewed (approved):
utACK 2a746500fa769f83405349bcffa3f0971b92a2de

Left a comment about usage of openmp, though I don't think it's a problem.

Out of curiosity though, is it only disabled to safe on compilation time and size? Or does it avoid some problem?

sedited reviewed (approved):
ACK 2a746500fa769f83405349bcffa3f0971b92a2de
